### PR TITLE
feat: Add ratings to insights

### LINKED ISSUE #14634: [CAL-3487] show CSAT in /insights
Add CSAT to insights

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3487](https://linear.app/calcom/issue/CAL-3487/csat)</sub>

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

This PR adds rating feedback metrics to insights. Following are the screenshots with various metrics added.

### There are three overview cards added in a new panel; No Show, Average Ratings & CSAT Score

- No Show represents the number of Host No Shows during the period
- Average Ratings represents the average ratings received by everyone in the group
- CSAT Score represents the Customer Satisfaction Score. It is calculated as follows:
    image
- No Show Host is added in the timeline to keep track of it over a time period
_______________________________________
### New tables have been added

- **Recent ratings**: Contains the recent bookings which were rated by the attendee, with host name & avatar, the actual rating and any feedback/notes. This table only contains ratings which actually had a note/feedback provided alongside of the ratings, sorted by most recent first
- **Most no show members**: Contains a list of hosts with no shows, maximum number of no shows first, with host name & avatar. The number of no shows are a collective per host.
- **Members with highest rated meetings**: Contains average rating received by individual hosts, sorted from highest rating to lowest, with host name & avatar along with the average rating
- **Members with lowest rated meetings**: Contains average rating received by individual hosts, sorted from lowest rating to highest, with host name & avatar along with the average rating

image

_______________________________________

**Here's how all of it looks:**

1. Primary cards with NoShow, AverageRatings & CSAT
image
________________________________________

2. NoShowHost in the timeline
image
________________________________________

3. Recent ratings, Most no show members, Members with highest rated meetings, Members with lowest rated meetings
image
________________________________________

### Added fields to CSV download:

- rating,
- ratingFeedback
- noShowHost

Fixes #14634 

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is a UI/UX design document, please, share it here.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How should this be tested?

Simply create a "Ratings" workflow and provide rating feedback once the events complete. You should see the ratings, feedback, noShow & CSAT in the insights page.

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: packages/features/insights/components/HighestNoShowHostTable.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (!isSuccess || !startDate || !endDate || !teamId) return null;
+
+  return data && data.length > 0 ? (
+    <CardInsights className="shadow-none">
+      <Title className="text-emphasis">{t("most_no_show_host")}</Title>
+      <TotalUserFeedbackTable data={data} />
+    </CardInsights>
+  ) : (
+    <></>
```
Comment (alishaz-polymath) [inline packages/features/insights/components/HighestNoShowHostTable.tsx:45]: We only show it if there's actual useful data here

File: packages/features/insights/components/HighestRatedMembersTable.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (!isSuccess || !startDate || !endDate || !teamId) return null;
+
+  return data && data.length > 0 ? (
+    <CardInsights className="shadow-none">
+      <Title className="text-emphasis">{t("highest_rated_members")}</Title>
+      <TotalUserFeedbackTable data={data} />
+    </CardInsights>
+  ) : (
+    <></>
```
Comment (alishaz-polymath) [inline packages/features/insights/components/HighestRatedMembersTable.tsx:45]: Again, show the table only if there's actual data in here

File: packages/features/insights/components/LowestRatedMembersTable.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (!isSuccess || !startDate || !endDate || !teamId) return null;
+
+  return data && data.length > 0 ? (
+    <CardInsights className="shadow-none">
+      <Title className="text-emphasis">{t("lowest_rated_members")}</Title>
+      <TotalUserFeedbackTable data={data} />
+    </CardInsights>
+  ) : (
+    <></>
```
Comment (alishaz-polymath) [inline packages/features/insights/components/LowestRatedMembersTable.tsx:45]: Same as above, show only when there's useful data in here

File: packages/features/insights/components/FeedbackTable.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                </TableCell>
+                <TableCell className="text-right">
+                  <Text>
+                    <strong className="text-default">{item.feedback}</strong>
+                  </Text>
+                </TableCell>
+              </TableRow>
+            ))
+          ) : (
+            <EmptyScreen Icon="zap" headline={t("no_ratings")} description={t("no_ratings_description")} />
```
Comment (alishaz-polymath) [inline packages/features/insights/components/FeedbackTable.tsx:54]: We use this empty screen if there are no ratings to let users know what to do to enable this

File: packages/prisma/migrations/20240419114622_add_ratings_to_insights/migration.sql [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    END AS "timeStatus",
+    et."parentId" AS "eventParentId",
+    "u"."email" AS "userEmail",
+    "u"."username" AS "username",
+    "Booking"."ratingFeedback",
+    "Booking"."rating",
+    "Booking"."noShowHost"
+FROM "Booking"
+LEFT JOIN "EventType" et ON "Booking"."eventTypeId" = et.id
+LEFT JOIN users u ON u.id = "Booking"."userId";
```
Comment (alishaz-polymath) [inline packages/prisma/migrations/20240419114622_add_ratings_to_insights/migration.sql:34]: Can only create or replace a view, not edit it currently

File: packages/prisma/schema.prisma [hunk 6]
```diff
@@ -1121,25 +1121,28 @@ enum AccessScope {
...
+  userId         Int?
+  teamId         Int?
+  eventLength    Int?
+  timeStatus     String?
+  eventParentId  Int?
+  userEmail      String?
+  username       String?
+  ratingFeedback String?
+  rating         Int?
+  noShowHost     Boolean?
```
Comment (alishaz-polymath) [inline packages/prisma/schema.prisma:1145]: 3 new columns added for rating and noshow

File: packages/features/insights/server/trpc-router.ts [hunk 7]
```diff
@@ -112,6 +112,18 @@ const emptyResponseEventsByStatus = {
...
+    deltaPrevious: 0,
+  },
+  no_show: {
+    count: 0,
+    deltaPrevious: 0,
+  },
+  csat: {
+    count: 0,
+    deltaPrevious: 0,
+  },
```
Comment (alishaz-polymath) [inline packages/features/insights/server/trpc-router.ts:126]: defaults

**Review by alishaz-polymath — COMMENTED** [review]
Self review added. It's all self explanatory and easy to follow tbh

File: packages/features/insights/components/HighestNoShowHostTable.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      teamId,
+      eventTypeId: selectedEventTypeId ?? undefined,
+      isAll,
+    },
+    {
+      staleTime: 30000,
+      trpc: {
+        context: { skipBatch: true },
+      },
+      enabled: !!(initialConfig?.teamId || initialConfig?.userId || initialConfig?.isAll),
```
Comment (emrysal) [inline packages/features/insights/components/HighestNoShowHostTable.tsx:31]: The earlier de-structuring makes me suspect this doesn't work, needs clarification.

**Review by emrysal — DISMISSED** [review]
Looks 👍 , just implement getUserAvatar and good to go.

**Review by Udit-takkar — COMMENTED** [review]
@alishaz-polymath how should we handle this case?

 In a collective team event type where multiple team members are fixed hosts. Rating is given to only one host
 
Screenshot 2024-04-22 at 2 39 20 PM

File: packages/features/insights/server/trpc-router.ts [hunk 9]
```diff
@@ -1437,6 +1488,538 @@ export const insightsRouter = router({
...
+        select: {
+          userId: true,
+          rating: true,
+          ratingFeedback: true,
+        },
+        take: 10,
+      });
+
+      const userIds = bookingsFromTeam.reduce((userIds: number[], booking) => {
+        if (typeof booking.userId === "number" && !userIds.includes(booking.userId)) {
```
Comment (Udit-takkar) [inline packages/features/insights/server/trpc-router.ts:1592]: NIT:
```suggestion
        if (!!booking.userId && !userIds.includes(booking.userId)) {
```

**Comment by alishaz-polymath** [comment]
> @alishaz-polymath how should we handle this case?
> 
> In a collective team event type where multiple team members are fixed hosts. Rating is given to only one host
> Screenshot 2024-04-22 at 2 39 20 PM

I was wondering about that this morning. I think the best way would be to just forward that to each host I think, so a collective event marked well should reflect the same rating to each individual, and vice-versa. Another thing I was thinking of was, multiple attendees, each attendee giving a new feedback. Currently, we would simply overwrite it. So I think we need a follow up to handle more than one hosts & more than one attendees logic, with proper RFC.

**Review by emrysal — DISMISSED** [review]
lgtm

File: packages/features/insights/components/FeedbackTable.tsx [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
```
Comment (PeerRich) [inline packages/features/insights/components/FeedbackTable.tsx:55]: NIT: maybe a button that links to "/workflows" and a "learn more" with a link to docs how to set up ratings

**Review by PeerRich — DISMISSED** [review]

**Comment by keithwillcode** [comment]
@alishaz-polymath how did this perform locally when having lots of insights data?

**Comment by alishaz-polymath** [comment]
> Instead of NoShowHost lets call this "No-Show (Host)"

Where specifically? All user facing elements?

**Comment by alishaz-polymath** [comment]
> @alishaz-polymath how did this perform locally when having lots of insights data?

Haven't tested against that yet.

**Comment by alishaz-polymath** [comment]
> > > @alishaz-polymath how did this perform locally when having lots of insights data?
> > 
> > 
> > Haven't tested against that yet.
> 
> Ok reminder there’s a seed script to populate lots of data for insights testing.

I wasn't aware of it, will test 🙏
Thank you!!

**Comment by alishaz-polymath** [comment]
> @alishaz-polymath how did this perform locally when having lots of insights data?

@keithwillcode Looks good


I have also updated the script to include new data

**Review by PeerRich — COMMENTED** [review]
great work. excited for this

File: packages/features/insights/components/TotalUserFeedbackTable.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+}: {
+  data:
+    | {
+        userId: number | null;
+        user: Pick<User, "avatarUrl" | "name">;
+        emailMd5?: string;
+        count?: number;
+        averageRating?: number | null;
+        username?: string;
+      }[]
```
Comment (sean-brydon) [inline packages/features/insights/components/TotalUserFeedbackTable.tsx:18]: NIT lets extract this type - its used elsewhere too

File: packages/features/insights/components/FeedbackTable.tsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+}: {
+  data:
+    | {
+        userId: number | null;
+        user: Pick<User, "avatarUrl" | "name">;
+        emailMd5?: string;
+        username?: string;
+        rating: number | null;
+        feedback: string | null;
+      }[]
```
Comment (sean-brydon) [inline packages/features/insights/components/FeedbackTable.tsx:19]: nit: Same here

File: packages/features/insights/components/LowestRatedMembersTable.tsx [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      },
+      enabled: !!(initialConfig?.teamId || initialConfig?.userId || initialConfig?.isAll),
+    }
+  );
+
+  if (isPending) return <LoadingInsight />;
+
+  if (!isSuccess || !startDate || !endDate || !teamId) return null;
+
+  return data && data.length > 0 ? (
```
Comment (sean-brydon) [inline packages/features/insights/components/LowestRatedMembersTable.tsx:39]: Shall we early return so we dont need this tenery ?

**Review by sean-brydon — APPROVED** [review]
Tested locally - works well. Nice job! 

Few NITS but nothing blocking

**Pull request merged**